src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from typing import cast
+
+from pants.backend.python.util_rules.pex import VenvPex, PexRequest, PexRequirements, VenvPexProcess
+from pants.backend.python.util_rules.interpreter_constraints import InterpreterConstraints
+from pants.backend.python.target_types import ConsoleScript, PythonRequirementsField
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.engine.fs import Digest, CreateDigest, FileContent, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import ProcessResult
+from pants.engine.rules import goal_rule, Get, collect_rules, MultiGet
+from pants.engine.target import Targets
+from pants.util.strutil import pluralize
+
+
+logger = logging.getLogger(__name__)
+
+
+class PipCompile(PythonToolBase):
+    options_scope = "pip_compile"
+    help = "Used to generate lockfiles for third-party Python dependencies."
+
+    default_version = "pip-tools==6.2.0"
+    default_main = ConsoleScript("pip-compile")
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        # TODO: How should users indicate where to save the lockfile to when we have per-tool
+        #  lockfiles and mmultiple user lockfiles?
+        register(
+            "--lockfile-dest",
+            type=str,
+            default="3rdparty/python/lockfile.txt",
+            help="The file path to be created.\n\nThis will overwrite any previous files."
+        )
+
+    @property
+    def lockfile_dest(self) -> str:
+        return cast(str, self.options.lockfile_dest)
+
+
+class LockSubsystem(GoalSubsystem):
+    name = "lock"
+
+
+class LockGoal(Goal):
+    subsystem_cls = LockSubsystem
+
+
+# TODO: What should the input to `./pants lock` be when generating user lockfiles? Currently, it's
+#  the 3rd-party requirements that get used as input. Likely it should instead be all 3rd party
+#  reqs used by the transitive closure? That would better mirror generate_lockfile.sh.
```

Comment:
Pretty sure the comment's proposed semantics are more correct

Reply:
Yes, I tend to agree.

Reply:
+1

This probably doesn't need to be exposed as a goal at all (in this PR)...?

Reply:
It's helpful for me to have a goal to run when iterating. But feature gated behind the experimental backend